swirl_lm/boundary_condition/immersed_boundary_method.py

```diff
@@ -1124,6 +1699,252 @@ def _apply_feedback_force_1d_interp(
+          p = lower_left_index[0] + i
+          q = lower_left_index[1] + j
+          s = lower_left_index[2] + k
+
+          # Make sure val is a tf.Tensor as `common_ops.gather` is not
+          # compatible with lists.
+          if not isinstance(val, tf.Tensor):
+            val = tf.convert_to_tensor(val, dtype=tf.float32)
+
+          #@NOTE: val is in z-x-y -> s-p-q!
+          res.append(common_ops.gather(val, tf.stack([s, p, q], axis=1)))
+
+        return res
+
+      # Gather the flow field values of `u` at the grid nodes enclosing each
+      # image point.
+      val_pt = get_corner_points(u, [idx_p, idx_q, idx_s])
+
+      # Adding additional value of the target quantity at the body intercept if
+      # boundary condition is a Dirichlet condition. The body intercept value is
+      # excluded for Neumann boundary conditions.
+      if sign == -1:
+        val_pt.append(tf.math.multiply(tf.ones_like(dist), bi_value))
+        weights = interp_weights
+      else:
+        weights = interp_weights_neumann
+
+      val_pt = tf.convert_to_tensor(val_pt, dtype=tf.float32)
+
+      # Multiply each grid node of the enclosing cell with its corresponding
+      # interpolation weight.
+      weighted_vals = tf.nest.map_structure(tf.math.multiply, weights, val_pt)
+
+      # Compute the interpolated flow field value at the image point
+      interped_val = tf.math.reduce_sum(weighted_vals, axis=0)
+
+      # Compute the ghost point value depending on the desired boundary
+      # condition
+      if sign == -1:
+        # DIRICHLET BOUNDARY CONDITION
+        gp_val = tf.nest.map_structure(
+            lambda _u: sign * _u + 2 * bi_value, interped_val
+        )
+      else:
+        # NEUMANN BOUNDARY CONDITION
+        # @NOTE: This approach is only first order accurate! See Luo et al.,
+        # International Journal of Heat and Mass Transfer 92, 2016, p. 710.
+        logging.warn(
+            'Neumann boundary condition assumes zero-gradient. '
+            'This approach is first order accurate only!'
+        )
+        gp_val = interped_val
+
+      # Since the ghost point values are stored in a 1D-vector, we need to
+      # scatter these values to the corresponding grid node in the 3D grid,
+      # i.e., generate a 3D tensor.
+      # gp_tensor will only have non-zero values at ghost point indices
+      # (ijk_gp), all other points will be zero.
+      gp_tensor = common_ops.scatter(
+          gp_val,
+          ijk_gp,
+          shape=[
+              self._params.core_nz + 2 * self._params.halo_width,
+              self._params.core_nx + 2 * self._params.halo_width,
+              self._params.core_ny + 2 * self._params.halo_width
+          ],
+          dtype=tf.float32
+      )
+
+      return tf.nest.map_structure(
+          lambda _u, _u_F: -beta * (_u - _u_F), u, gp_tensor
+      )
+
+    return get_ib_force
+
+  def _apply_ghost_point_method(
```

Comment:
I think we should be able to consolidate this function with the other direct forcing methods by replacing the force function.